### ISSUE TITLE
How to add screenshots from local to serenity report for each step or on failures

### ISSUE THREAD

**Issue by vdevarapu** [serenity-bdd/serenity-documentation#109]
Could you please let me know on how to add screenshots from local folder to Serenity reports.

**Comment by wakaleo** [comment]
Why do you need to add screenshots from a local folder?

**Comment by vdevarapu** [comment]
Thank you so much for the reply.

We are trying to implement Serenity for SeeTest code in seetest the driver is MobileWebDriver.

I believe we do not have MobileWebDriver in pageObject class and hence we are not able to get the driver instance. So we are creating driver object from my local and we are running the scripts against IOS mobile devices

driver = new MobileWebDriver( host, port);

Thanks
--------------------------------------------
On Tue, 20/6/17, John Ferguson Smart <notifications@github.com> wrote:

 Subject: Re: [serenity-bdd/serenity-documentation] How to add screenshots from local to serenity report for each step or on failures (#109)
 To: "serenity-bdd/serenity-documentation" <serenity-documentation@noreply.github.com>
 Cc: "vdevarapu" <venkat_220@yahoo.co.in>, "Author" <author@noreply.github.com>
 Date: Tuesday, 20 June, 2017, 4:10 PM
 
 Why do you need to add
 screenshots from a local folder?
 
 —
 You are
 receiving this because you authored the thread.
 Reply
 to this email directly, view
 it on GitHub, or mute
 the thread.

**Comment by wakaleo** [comment]
Can you not use a ProvidedDriver that creates a new MobileWebDriver instance instead? Does the new MobileWebDriver implement the WebDriver interface?

**Comment by vdevarapu** [comment]
Yes MobileWebDriver implements webdriver, and I am not sure on how to use ProvidedDriver.

**Comment by vdevarapu** [comment]
And also there are few specific methods in MobileWebDriver like setDevice(), will we be able to get these methods if we create a ProvidedDriver and kindly help me in getting the reports if ProvidedDriver is implemented.

**Comment by wakaleo** [comment]
I will write up some documentation on how to do it when I get a chance, but there are plenty in the Serenity Google Group who use the ProvidedDriver, so maybe ask there for some examples as well.

**Comment by vdevarapu** [comment]
Sure, will do that, could you please let me know if serenity will be feasible for MobileWebDriver.

**Comment by wakaleo** [comment]
I see no reason why it wouldn't work if it implements WebDriver.

**Comment by vdevarapu** [comment]
I did this in my code,

public class loginPage extends PageObject implements DriverSource{
    
    private String host = "localhost";
    private int port = 8889;

    MobileWebDriver driver = null;

    @Override
    public WebDriver newDriver() {
        return new MobileWebDriver(host, port);
    }

    @Override
    public boolean takesScreenshots() {
        return true;
    }

**Comment by vdevarapu** [comment]
Hi please find the code that I implemented but when I run this I am not able to get the driver instance and the test fails.

public class loginPage extends PageObject implements DriverSource{

    private String host = "localhost";
    private int port = 8889;

    MobileWebDriver driver = null;

    @Override
    public WebDriver newDriver() {

        driver = new MobileWebDriver(host, port);
        return driver;
    }

    @Override
    public boolean takesScreenshots() {
        return true;
    }

    public void callGivenPage() throws Exception
    {
        System.out.println("In the Given block");
        driver = (MobileWebDriver)((WebDriverFacade)getDriver()).getProxiedDriver();
        driver.setDevice("ios_app:Fnu's iPhone");
        driver.useNativeIdentification();
        driver.findElement(By.xpath("/*//*[@placeholder='Employee ID']")).click();

**Comment by vdevarapu** [comment]
I provided the following information in serenity.properties file and when I run with the above configuration, I am getting the error "net.thucydides.core.webdriver.UnsupportedDriverException: Could not instantiate class net.thucydides.core.webdriver.ProvidedDriver". Please help.

webdriver.driver = provided
webdriver.provided.type = mydriver
webdriver.provided.mydriver = com.experitest.selenium.MobileWebDriver
com.experitest.selenium.MobileWebDriver.port=8889
com.experitest.selenium.MobileWebDriver.host=localhost
thucydides.driver.capabilities = mydriver

**Comment by vdevarapu** [comment]
Hi John, it would be great if you can please let me know if there is any issue with the code I wrote or is it something Serenity team needs to handle so that our team will not spend much time debugging this as they spent on attaching screenshots from local. Thanks and awaiting for your response.

**Comment by vdevarapu** [comment]
Hi John, Please let me know if you have any update.

**Comment by wakaleo** [comment]
Your LoginPage should have no changes or overrides (i.e. none of the code you mention above). All of this logic goes in a separate ProvidedDriver class (separation of concerns). Have a look at https://thucydides-webtests.com/2014/03/07/thucydides-version-0-9-235-released/ for some information on how to implement a ProvidedDriver class.
